### PR TITLE
Kernel module

### PR DESCRIPTION
**Pull request by FunFunFine**
This PR extracts every typeclass without dependency on `cats-effect` from the `core` module into the new `kernel` one. It is cats-effect free as that was the point.

Also, I split `Context.scala` a bit.

Nothing is removed.

Although, there is a problem with splitting tofu package object. The current solution does not seem to work well.

Merging this PR allows us to release a cats-effect3-compatible version of a part of tofu.

### CODE REVIEW

**Review by Z1kkurat — COMMENTED** [review]
To sum up, this is the new location of user typeclasses and friends (probably will be useful as some kind of migration guide). Please feel free to add and/or correct if I missed anything.

`common`:
- Fire/Race/Start and its syntax (`tofu.syntax.fire`)
- Timeout and its syntax (`tofu.syntax.timeout`)
- Console and its syntax (`tofu.syntax.console`)
- time utilities: TimeZone, TimeData and all the corresponding syntax (`tofu.syntax.time`)
- GenRandom
- GenUUID
- `tofu.syntax.timer`

`kernel`:
- context-related typeclasses, aliases and syntax:`Context`, `WithContext`, `HasContext`, `Local`, `HasLocal`, `WithLocal`, `Provide`, `HasProvide`, `WithProvide`, `RunContext`, `WithRun`, `HasContextRun`
- error-related typeclasses, aliases and syntax: `Raise`, `RestoreTo`, `Restore`, `HandleTo`, `Handle`, `ErrorsTo`
- `tofu.Init`
- `tofu.Void`
- bifunctor-related typeclasses and aliases: `tofu.bi.BiContext`, `tofu.bi.BiLocal`, `tofu.bi.BiRun`, `tofu.bi.lift.BiLift`, `tofu.bi.lift.BiUnlift`
- `tofu.control`: ApplyZip, ApplicativeZip, ContravariantFilterConsume, Optional, Partial, Selective, Switch
- `tofu.lift`, aliases and syntax: IsoK, Rebase, Lift, Unlift
- `tofu.streaming`: Push, Pull
- `tofu.syntax.etiher` / `tofu.syntax.feither`
- `tofu.syntax.foldable`
- `tofu.syntax.foption`
- `tofu.syntax.monadError`
- `tofu.syntax.monoid`
- `tofu.syntax.optics`
- `tofu.syntax.selective`
- `tofu.syntax.traverse`

File: project/Dependencies.scala [hunk 1]
```diff
@@ -16,7 +16,8 @@ object Dependencies {
 
     val cats = "2.6.1"
 
-    val catsEffect = "2.5.1"
+    val catsEffect  = "2.5.1"
+    val catsEffect3 = "3.1.1"
```
Comment (Odomontois) [inline project/Dependencies.scala:20]: I think it's better not to mention cats effect 3 until real start of migration

File: build.sbt [hunk 2]
```diff
@@ -35,6 +35,12 @@ lazy val higherKindCore = project settings (
   libraryDependencies ++= Seq(catsCore, catsFree, catsTagless),
 )
 
+lazy val kernel = project in file("kernel") dependsOn (opticsCore, higherKindCore) settings (
```
Comment (Odomontois) [inline build.sbt:38]: What I feel to to best here is to make `core` dependent on `kernel`, moving desired type classes instead of copying them

Comment (FunFunFine) [reply build.sbt:37]: Done

File: build.sbt [hunk 3]
```diff
@@ -34,8 +34,12 @@ lazy val higherKindCore = project settings (
...
   libraryDependencies ++= Seq(catsCore, catsFree, catsTagless),
 )
-
-lazy val core = project dependsOn (opticsCore, higherKindCore) settings (
+lazy val kernel = project in file("kernel") dependsOn (opticsCore, higherKindCore) settings (
+  defaultSettings,
+  publishName := "kernel",
+  libraryDependencies ++= Seq(catsCore, catsTagless),
+)
+lazy val core = project dependsOn (opticsCore, higherKindCore, kernel) settings (
```
Comment (Z1kkurat) [inline build.sbt:42]: Do we need to depend on `opticsCore` and `higherKindCore` since `kernel` already depends on them?

**Review by catostrophe — COMMENTED** [review]
Will check later

File: build.sbt [hunk 4]
```diff
@@ -319,7 +324,7 @@ lazy val defaultScalacOptions = scalacOptions := {
   if (!sys.env.get("CI").contains("true") || (minorVersion.value == 12))
     opts.filterNot(Set("-Xfatal-warnings"))
   else
-    opts
+    opts ++ Seq("-Xlog-implicits")
```
Comment (Odomontois) [inline build.sbt:327]: это что

Comment (FunFunFine) [reply build.sbt:327]: sure
```suggestion
    opts
```

**Comment by FunFunFine** [comment]
closed in favor of #675

**Pull request closed without merging**